062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+It will be inside the `.spec.kafka` section.
+And it will allow configuring different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The build-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - the Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
+* The other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quotas plugin - will be configured as `type: strimzi`.
+The API will allow to configure the storage limits at which the plugin will start limiting the producers in order to prevent the disk from getting full.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: strimzi
+      produceQuota: 1000000
+      fetchQuota: 1000000
+      minAvailableBytesPerVolume: 5368709120
+      excludedPrincipals:
+        - my-special-user
+    # ... 
+  # ...
+```
+
+The `type: strimzi` quotas API will support the following fields:
+* `produceQuota` for setting the default produce quota
+* `fetchQuota` for setting the default fetch quota
+* `minAvailableBytesPerVolume` for setting the minimal required free space per volume as number of bytes (this field can be configured only when `minAvailableRatioPerVolume`is not set)
+* `minAvailableRatioPerVolume` for setting the minimal required free space per volume as percentage of the overall volume size (this field can be configured only when `minAvailableBytesPerVolume`is not set)
+* `excludedPrincipals` for setting s list of users to be excluded from the applied quotas
+
+Other options might be configured directly by the user in `.spec.kafka.config` section of the `Kafka` CR.
+
+Based on this configuration, the operator will:
+* Set the `client.quota.callback.class` field in Kafka configuration to `io.strimzi.kafka.quotas.StaticQuotaCallback`
```

Comment:
It seem inherent in the proposal but I think it would help to state it explicitly. Defining `spec.kafka.quotas` implicitly overrides any value specified for `client.quota.callback.class` in `kafka.config`

Reply:
I think this is explained on  the current lines 147-148?

Reply:
Yes, I'd agree. I didn't pick it up and remove this when reading sequentially initially.